### PR TITLE
PySpaces integration

### PR DESCRIPTION
**Pull request by cbensimon**
Allow Spaces infra to intercept Gradio backend handlers

Internal Slack conversation :
https://huggingface.slack.com/archives/C02Q9UFS8AW/p1685448898402919

### CODE REVIEW

File: test/requirements.in [hunk 1]
```diff
@@ -16,6 +16,7 @@ ruff>=0.0.260
 respx
 scikit-image
 shap
+spaces; python_version >= '3.9'
```
Comment (abidlabs) [inline test/requirements.in:19]: Why do we add `spaces` to the test requirements when it is not actually used in any tests?

File: gradio/blocks.py [hunk 2]
```diff
@@ -388,6 +393,18 @@ def __init__(
...
+
+    def spaces_auto_wrap(self):
+        if spaces is None:
+            return
+        if not utils.is_space():
+            return
+        wrapped_fn = spaces.gradio_auto_wrap(self.fn)
+        if wrapped_fn is self.fn:
+            return
+        print(f"Your `{self.name}` event handler has been optimized for Spaces")
```
Comment (abidlabs) [inline gradio/blocks.py:406]: I don't think we need to actually print this (can get noisy for Blocks with multiple functions), but instead I meant we can include some messaging in the changelog saying that "Functions in Gradio apps will automatically get optimized for Spaces, and <here's how to disable that>"

File: test/requirements.in (hunk 1 above)
Comment (cbensimon) [reply test/requirements.in:19]: It is needed so typechecking (pyright) doesn't error out

I can still `# type: ignore` the problematic line instead if you are trying to keep the test requirements as small as possible

File: gradio/blocks.py (hunk 2 above)
Comment (cbensimon) [reply gradio/blocks.py:406]: Ok I totally missed your point then ! I'll add the line in the changelog

File: test/requirements.in (hunk 1 above)
Comment (abidlabs) [reply test/requirements.in:19]: Yup that'd be better thanks!

File: gradio/blocks.py (hunk 2 above)
Comment (abidlabs) [reply gradio/blocks.py:406]: Sorry for the confusion!

**Comment by cbensimon** [comment]
> Thanks @cbensimon just a couple of clarifications above! Would you be able to add an entry in `CHANGELOG.md` with a description of the change as well? Thanks

Done !

**Comment by abidlabs** [comment]
Merged in, thanks so much @cbensimon!

**Pull request merged**